File: vi/codes/swift/chapter_graph/graph_adjacency_matrix.swift
/**
 * File: graph_adjacency_matrix.swift
 * Created Time: 2023-02-01
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

/* Lớp đồ thị vô hướng dựa trên ma trận kề */
class GraphAdjMat {
    private var vertices: [Int] // Danh sách đỉnh, trong đó phần tử biểu diễn "giá trị đỉnh" và chỉ số biểu diễn "chỉ số đỉnh"
    private var adjMat: [[Int]] // Ma trận kề, trong đó chỉ số hàng và cột tương ứng với "chỉ số đỉnh"

    /* Hàm khởi tạo */
    init(vertices: [Int], edges: [[Int]]) {
        self.vertices = []
        adjMat = []
        // Thêm đỉnh
        for val in vertices {
            addVertex(val: val)
        }
        // Thêm cạnh
        // Lưu ý các phần tử của edges biểu diễn chỉ số đỉnh, tức tương ứng với chỉ số các phần tử của vertices
        for e in edges {
            addEdge(i: e[0], j: e[1])
        }
    }

    /* Lấy số lượng đỉnh */
    func size() -> Int {
        vertices.count
    }

    /* Thêm đỉnh */
    func addVertex(val: Int) {
        let n = size()
        // Thêm giá trị của đỉnh mới vào danh sách đỉnh
        vertices.append(val)
        // Thêm một hàng vào ma trận kề
        let newRow = Array(repeating: 0, count: n)
        adjMat.append(newRow)
        // Thêm một cột vào ma trận kề
        for i in adjMat.indices {
            adjMat[i].append(0)
        }
    }

    /* Xóa đỉnh */
    func removeVertex(index: Int) {
        if index >= size() {
            fatalError("Out of bounds")
        }
        // Xóa đỉnh tại vị trí index khỏi danh sách đỉnh
        vertices.remove(at: index)
        // Xóa hàng tại vị trí index khỏi ma trận kề
        adjMat.remove(at: index)
        // Xóa cột tại vị trí index khỏi ma trận kề
        for i in adjMat.indices {
            adjMat[i].remove(at: index)
        }
    }

    /* Thêm cạnh */
    // Tham số i, j tương ứng với chỉ số các phần tử của vertices
    func addEdge(i: Int, j: Int) {
        // Xử lý trường hợp chỉ số vượt giới hạn và bằng nhau
        if i < 0 || j < 0 || i >= size() || j >= size() || i == j {
            fatalError("Out of bounds")
        }
        // Trong đồ thị vô hướng, ma trận kề đối xứng qua đường chéo chính, tức (i, j) == (j, i)
        adjMat[i][j] = 1
        adjMat[j][i] = 1
    }

    /* Xóa cạnh */
    // Tham số i, j tương ứng với chỉ số các phần tử của vertices
    func removeEdge(i: Int, j: Int) {
        // Xử lý trường hợp chỉ số vượt giới hạn và bằng nhau
        if i < 0 || j < 0 || i >= size() || j >= size() || i == j {
            fatalError("Out of bounds")
        }
        adjMat[i][j] = 0
        adjMat[j][i] = 0
    }

    /* In ma trận kề */
    func print() {
        Swift.print("Vertex list = ", terminator: "")
        Swift.print(vertices)
        Swift.print("Adjacency matrix =")
        PrintUtil.printMatrix(matrix: adjMat)
    }
}

@main
enum GraphAdjacencyMatrix {
    /* Mã điều khiển */
    static func main() {
        /* Thêm cạnh */
        // Lưu ý các phần tử của edges biểu diễn chỉ số đỉnh, tức tương ứng với chỉ số các phần tử của vertices
        let vertices = [1, 3, 2, 5, 4]
        let edges = [[0, 1], [1, 2], [2, 3], [0, 3], [2, 4], [3, 4]]
        let graph = GraphAdjMat(vertices: vertices, edges: edges)
        print("\nAfter initialization, graph is")
        graph.print()

        /* Thêm cạnh */
        // Thêm đỉnh
        graph.addEdge(i: 0, j: 2)
        print("\nAfter adding edge 1-2, graph is")
        graph.print()

        /* Xóa cạnh */
        // Đỉnh 1, 3 có chỉ số lần lượt là 0, 1
        graph.removeEdge(i: 0, j: 1)
        print("\nAfter removing edge 1-3, graph is")
        graph.print()

        /* Thêm đỉnh */
        graph.addVertex(val: 6)
        print("\nAfter adding vertex 6, graph is")
        graph.print()

        /* Xóa đỉnh */
        // Đỉnh 3 có chỉ số 1
        graph.removeVertex(index: 1)
        print("\nAfter removing vertex 3, graph is")
        graph.print()
    }
}

File: vi/codes/csharp/chapter_graph/graph_adjacency_matrix.cs
/**
 * File: graph_adjacency_matrix.cs
 * Created Time: 2023-02-06
 * Author: zjkung1123 (zjkung1123@gmail.com)
 */

namespace hello_algo.chapter_graph;

/* Lớp đồ thị vô hướng dựa trên ma trận kề */
class GraphAdjMat {
    List<int> vertices;     // Danh sách đỉnh, trong đó phần tử biểu thị "giá trị đỉnh" và chỉ số biểu thị "chỉ số đỉnh"
    List<List<int>> adjMat; // Ma trận kề, trong đó chỉ số hàng và cột tương ứng với "chỉ số đỉnh"

    /* Hàm khởi tạo */
    public GraphAdjMat(int[] vertices, int[][] edges) {
        this.vertices = [];
        this.adjMat = [];
        // Thêm đỉnh
        foreach (int val in vertices) {
            AddVertex(val);
        }
        // Thêm cạnh
        // Lưu ý rằng các phần tử của edges biểu thị chỉ số đỉnh, tức tương ứng với chỉ số phần tử của vertices
        foreach (int[] e in edges) {
            AddEdge(e[0], e[1]);
        }
    }

    /* Lấy số lượng đỉnh */
    int Size() {
        return vertices.Count;
    }

    /* Thêm đỉnh */
    public void AddVertex(int val) {
        int n = Size();
        // Thêm giá trị của đỉnh mới vào danh sách đỉnh
        vertices.Add(val);
        // Thêm một hàng vào ma trận kề
        List<int> newRow = new(n);
        for (int j = 0; j < n; j++) {
            newRow.Add(0);
        }
        adjMat.Add(newRow);
        // Thêm một cột vào ma trận kề
        foreach (List<int> row in adjMat) {
            row.Add(0);
        }
    }

    /* Xóa đỉnh */
    public void RemoveVertex(int index) {
        if (index >= Size())
            throw new IndexOutOfRangeException();
        // Xóa đỉnh tại index khỏi danh sách đỉnh
        vertices.RemoveAt(index);
        // Xóa hàng tại index khỏi ma trận kề
        adjMat.RemoveAt(index);
        // Xóa cột tại index khỏi ma trận kề
        foreach (List<int> row in adjMat) {
            row.RemoveAt(index);
        }
    }

    /* Thêm cạnh */
    // Các tham số i, j tương ứng với chỉ số phần tử của vertices
    public void AddEdge(int i, int j) {
        // Xử lý chỉ số vượt biên và trường hợp bằng nhau
        if (i < 0 || j < 0 || i >= Size() || j >= Size() || i == j)
            throw new IndexOutOfRangeException();
        // Trong đồ thị vô hướng, ma trận kề đối xứng qua đường chéo chính, tức (i, j) == (j, i)
        adjMat[i][j] = 1;
        adjMat[j][i] = 1;
    }

    /* Xóa cạnh */
    // Các tham số i, j tương ứng với chỉ số phần tử của vertices
    public void RemoveEdge(int i, int j) {
        // Xử lý chỉ số vượt biên và trường hợp bằng nhau
        if (i < 0 || j < 0 || i >= Size() || j >= Size() || i == j)
            throw new IndexOutOfRangeException();
        adjMat[i][j] = 0;
        adjMat[j][i] = 0;
    }

    /* In ma trận kề */
    public void Print() {
        Console.Write("Vertex list = ");
        PrintUtil.PrintList(vertices);
        Console.WriteLine("Adjacency matrix =");
        PrintUtil.PrintMatrix(adjMat);
    }
}

public class graph_adjacency_matrix {
    [Test]
    public void Test() {
        /* Thêm cạnh */
        // Lưu ý rằng các phần tử của edges biểu thị chỉ số đỉnh, tức tương ứng với chỉ số phần tử của vertices
        int[] vertices = [1, 3, 2, 5, 4];
        int[][] edges =
        [
            [0, 1],
            [0, 3],
            [1, 2],
            [2, 3],
            [2, 4],
            [3, 4]
        ];
        GraphAdjMat graph = new(vertices, edges);
        Console.WriteLine("\nAfter initialization, graph is");
        graph.Print();

        /* Thêm cạnh */
        // Thêm đỉnh
        graph.AddEdge(0, 2);
        Console.WriteLine("\nAfter adding edge 1-2, graph is");
        graph.Print();

        /* Xóa cạnh */
        // Các đỉnh 1, 3 có chỉ số lần lượt là 0, 1
        graph.RemoveEdge(0, 1);
        Console.WriteLine("\nAfter removing edge 1-3, graph is");
        graph.Print();

        /* Thêm đỉnh */
        graph.AddVertex(6);
        Console.WriteLine("\nAfter adding vertex 6, graph is");
        graph.Print();

        /* Xóa đỉnh */
        // Đỉnh 3 có chỉ số 1
        graph.RemoveVertex(1);
        Console.WriteLine("\nAfter removing vertex 3, graph is");
        graph.Print();
    }
}
